core/domain/state_domain.py

```diff
@@ -194,7 +194,9 @@ def from_dict(cls, answer_group_dict: AnswerGroupDict) -> AnswerGroup:
     def validate(
         self,
         interaction: base.BaseInteraction,
-        exp_param_specs_dict: Dict[str, param_domain.ParamSpec]
+        exp_param_specs_dict: Dict[str, param_domain.ParamSpec],
+        *,
+        validation_from_exploration: bool = False
```

Comment:
I think we should avoid using exp/question term inside state class i.e, state class is generic! Can we call it tagged_skill_misconception_id_allowed?

Reply:
I saw an explanation here: https://github.com/oppia/oppia/pull/15797/files#diff-f02dad94abb3850a7cd43536e8251abdf5388d63f60a08f19c33f51e120d52fdR1902

I would suggest having two flags `tagged_skill_misconception_id_allowed` and `allow_empty_rule_specs` both defaults to False and in exp/question class call validate with required values and write code-comment in question class explaining why it's disabled. This will ensure that question class has all the information and it would be easier to do the cleanup later.  wdyt?

Reply:
Yes! I totally agree with you. I have made the changes, though I changed the variable name a bit. The reason is `allow_empty_rule_specs` kind of says that we allow empty rule specs which is not exactly true, it is only true for question if we provide training data instead. Please let me know if the changes looks good to you otherwise I will make the changes asap.
Thanks

Reply:
> I have made the changes, though I changed the variable name a bit

I agree `allow_empty_rule_specs` isn't a good name for the variable (As it starts with a verb), thanks for raising it!